tests/test_settings.py

```diff
@@ -275,6 +275,14 @@ class SettingsTestVector:
             custom_mapping_file={Path("mapping.toml")},
         ),
     ),
+    SettingsTestVector(
+        "config_file_with_pyenvs_and_cli__cli_pyenvs_overrides_config",
+        config=dict(pyenvs=["foo", "bar"]),
+        cmdline=dict(pyenvs=["baz", "xyzzy"]),
```

Comment:
Are we also defaulting to override config with cli arguments for other options that might have multiple values?

Reply:
Yes, here's what happens for e.g. `.code` (using environment vs. CLI, but the same holds for config file vs. CLI):
```sh
$ fawltydeps_code='["fawltydeps/main.py", "fawltydeps/packages.py"]' \
  fawltydeps --list-imports --detailed
fawltydeps/main.py:20: pydantic
fawltydeps/packages.py:20: importlib_metadata
fawltydeps/packages.py:37: tomli

$ fawltydeps_code='["fawltydeps/main.py", "fawltydeps/packages.py"]' \
  fawltydeps --list-imports --detailed --code fawltydeps/types.py fawltydeps/utils.py
fawltydeps/types.py:15: typing_extensions
fawltydeps/utils.py:9: importlib_metadata
```

When I first wrote the `Settings` class, I added the following TODO (and it's still there):
```
    TODO? Currently overrides happen _completely_, for example specifying
    --ignore-undeclared on the command-line will _replace_ an ignore_undeclared
    directive in pyproject.toml. We may want to consider allowing some
    directives to _combine_ (although that will carry further complications).
```

The "further complications" I allude to here are more complex than one might imagine at first:
- We would probably need to introspect over the `Settings` members and/or their type and define _different_ ways of combining them:
  - Most sets (`code`/`deps`/`pyenvs`/`custom_mapping_file`/`ignore_*`) combine by _union_
  - `custom_mapping` combines by _merging dicts_ (a la `accumulate_mappings()`)
  - How do we want `actions` to combine? One could argue for either _replace_ or _union_, but probably _replace_ makes most sense.
  - `output_format` should likely _replace_?
  - Same with `deps_parser_choice`?
  - The `verbosity`number should probably `sum()` across levels?
- When we decide to take the _union_ of multi-valued members, then how do we deal with wanting to _remove_ something that was added at an earlier stage? For example: the config file says `code = ["foo.py", "bar.py"]`, but I want FD to look at code from `bar.py` and `baz.py`; I can add `baz.py` at the command-line, but how do I _remove_ `foo.py`?
- We would maybe need to add more options (`--remove-code=foo.py` or `--no-code=foo.py`) or more complex option parsing (`--code "+baz.py" "-foo.py"`) to deal with this.
- How to deal with different options that "cooperate" in unexpected ways? E.g. `deps = ["pyproject.toml"]` in the config file, but `--deps my_reqs.pip --deps-parser-choice requirements.txt` on the command line would probably combine into the following nonsense in `Settings`:
    ```
    deps = {"pyproject.toml", "my_reqs.pip"}
    deps_parser_choice = ParserChoice.REQUIREMENTS_TXT
    ```

When I first wrote this code in #167, I decided to _not_ open this Pandora's box of complexity, and to leave everything as strictly overriding/replacing for now.

After revisiting it now, I still think this is the sanity-preserving choice! 😅

Reply:
I agree with that analysis. Could you post it in an issue as it's very useful and would otherwise be lost.

Reply:
https://github.com/tweag/FawltyDeps/issues/328 ✅